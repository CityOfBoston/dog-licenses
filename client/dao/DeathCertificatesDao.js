// @flow

import DataLoader from 'dataloader';
import type { LoopbackGraphql } from '../loopback-graphql';
import type { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import type { DeathCertificateSearchResults } from '../types';

export type DeathCertificateCache = {[id: string]: DeathCertificate};

export default class DeathCertificatesDao {
  loopbackGraphql: LoopbackGraphql;
  loader: DataLoader<string, ?DeathCertificate>;

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader((ids) => fetchDeathCertificates(loopbackGraphql, ids).then((a) => a.map((i) => i)));
  }

  get(id: string): Promise<?DeathCertificate> {
    return this.loader.load(id);
  }

  async search(query: string, page: number): Promise<DeathCertificateSearchResults> {
    const results = await searchDeathCertificates(this.loopbackGraphql, query, page);

    results.results.forEach((cert) => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }
}
