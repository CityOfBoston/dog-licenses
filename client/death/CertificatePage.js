// @flow

import React from 'react';
import { css } from 'glamor';
import Head from 'next/head';
import type { Context } from 'next';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import makeLoopbackGraphql from '../loopback-graphql';
import type { DeathCertificate } from '../types';

import type Cart from '../store/Cart';
import Nav from '../common/Nav';

const FORM_STYLE = css({
  display: 'flex',
  alignItems: 'center',
});

const QUANTITY_STYLE = css({
  minWidth: '5em',
});

const ADD_TO_CART_STYLE = css({
  flex: 1,
  marginLeft: '1em',
});

export type InitialProps = {|
  id: string,
  certificate: ?DeathCertificate,
|}

export type Props = {
  /* :: ...InitialProps, */
  cart: Cart,
}

type State = {
  quantity: number,
};

export default class CertificatePage extends React.Component {
  props: Props;
  state: State = {
    quantity: 1,
  }

  static async getInitialProps(ctx: Context<*>): Promise<InitialProps> {
    const { query: { id }, req } = ctx;

    const loopbackGraphql = makeLoopbackGraphql(req);
    const certificate = (await fetchDeathCertificates(loopbackGraphql, [id]))[0];

    return {
      id,
      certificate,
    };
  }

  handleQuantityChange = (ev: SyntheticInputEvent) => {
    this.setState({
      quantity: parseInt(ev.target.value, 10),
    });
  }

  handleAddToCart = (ev: SyntheticInputEvent) => {
    const { cart, certificate } = this.props;
    const { quantity } = this.state;

    ev.preventDefault();

    if (certificate) {
      cart.add(certificate, quantity);
    }
  }

  render() {
    const { id, certificate, cart } = this.props;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificate #{id}</title>
        </Head>

        <Nav cart={cart} link="checkout" />

        <div className="p-a300 b--g">
          <div className="sh sh--b0">
            <h1 className="sh-title" style={{ marginBottom: 0 }}>Deceased Details</h1>
          </div>
        </div>

        <div className="p-a300">
          { certificate && this.renderCertificate(certificate) }
        </div>
      </div>
    );
  }

  renderCertificate({ firstName, lastName, birthYear, deathYear, causeOfDeath }: DeathCertificate) {
    const { quantity } = this.state;

    return (
      <div>
        <ul className="dl">
          <li className="dl-i">
            <span className="dl-t">Full Name</span>
            <span className="dl-d">{firstName} {lastName}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Birth Year</span>
            <span className="dl-d">{birthYear}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Death Year</span>
            <span className="dl-d">{deathYear}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Cause of Death</span>
            <span className="dl-d">{causeOfDeath || 'Pending'}</span>
          </li>
        </ul>

        <form onSubmit={this.handleAddToCart} className={`js-add-to-cart-form m-v300 ${FORM_STYLE.toString()}`}>
          <select name="quantity" value={quantity} className={QUANTITY_STYLE} onChange={this.handleQuantityChange}>
            <option value="1">Qty: 1</option>
            <option value="2">Qty: 2</option>
            <option value="3">Qty: 3</option>
            <option value="4">Qty: 4</option>
            <option value="5">Qty: 5</option>
            <option value="6">Qty: 6</option>
            <option value="7">Qty: 7</option>
            <option value="8">Qty: 8</option>
            <option value="9">Qty: 9</option>
            <option value="10">Qty: 10</option>
          </select>

          <button type="submit" className={`btn ${ADD_TO_CART_STYLE.toString()}`}>Add to Cart</button>
        </form>
      </div>
    );
  }

}
