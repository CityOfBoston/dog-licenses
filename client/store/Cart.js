// @flow

import { observable, computed, action } from 'mobx';

import type { DeathCertificate } from '../types';

export class CartItem {
  @observable.ref cert: DeathCertificate;
  @observable quantity: number;
}

export default class Cart {
  @observable items = [];

  @computed get size(): number {
    return this.items.reduce((acc, item) => acc + item.quantity, 0);
  }

  @action add(cert: DeathCertificate, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === cert.id);
    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const item = new CartItem();
      item.cert = cert;
      item.quantity = quantity;

      this.items.push(item);
    }
  }

  @action setQuantity(certId: string, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === certId);
    if (existingItem) {
      existingItem.quantity = quantity;
    }
  }

  @action remove(certId: string) {
    const idx = this.items.findIndex(({ cert }) => cert.id === certId);
    if (idx !== -1) {
      this.items.splice(idx, 1);
    }
  }
}
