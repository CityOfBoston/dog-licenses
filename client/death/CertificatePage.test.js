// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import type { DeathCertificate } from '../types';
import Cart from '../store/Cart';
import DeathCertificatesDao from '../dao/DeathCertificatesDao';

import CertificatePage from './CertificatePage';
import type { InitialProps } from './CertificatePage';

jest.mock('../dao/DeathCertificatesDao');

const TEST_DEATH_CERTIFICATE: DeathCertificate = {
  id: '000002',
  firstName: 'Bruce',
  lastName: 'Banner',
  birthYear: '1962',
  deathYear: '2016',
  causeOfDeath: 'Hawkeye',
  age: '21 yrs.',
};

const renderFromInitialProps = async (query: {[key: string]: string}, dependencies: Object) => {
  const cart = new Cart();
  const initialProps: InitialProps = await CertificatePage.getInitialProps((({
    query,
  }): any), dependencies);

  return renderer.create(<CertificatePage cart={cart} {...initialProps} />);
};

describe('rendering', () => {
  let deathCertificatesDao;

  beforeEach(() => {
    deathCertificatesDao = new DeathCertificatesDao((null: any));
  });

  it('renders a certificate', async () => {
    deathCertificatesDao.get.mockReturnValue(TEST_DEATH_CERTIFICATE);
    expect((await renderFromInitialProps({ id: '000002' }, { deathCertificatesDao })).toJSON()).toMatchSnapshot();
    expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
  });

  it('renders a 404', async () => {
    deathCertificatesDao.get.mockReturnValue(null);

    expect((await renderFromInitialProps({ id: '000002' }, { deathCertificatesDao })).toJSON()).toMatchSnapshot();
    expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
  });
});

describe('searching', () => {
  it('redirects to search for a query', () => {
    const cart = new Cart();
    const wrapper = shallow(<CertificatePage cart={cart} id="00002" certificate={TEST_DEATH_CERTIFICATE} />);

    wrapper.find('select[name="quantity"]').simulate('change', { target: { value: '5' } });
    wrapper.find('form.js-add-to-cart-form').simulate('submit', { preventDefault: jest.fn() });

    expect(cart.size).toEqual(5);
  });
});
