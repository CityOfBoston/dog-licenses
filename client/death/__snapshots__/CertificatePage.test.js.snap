// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders a 404 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/death/checkout"
        onClick={[Function]}
      >
        Checkout
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/death/checkout"
        onClick={[Function]}
      >
        0
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0"
    >
      <h1
        className="sh-title"
        style={
          Object {
            "marginBottom": 0,
          }
        }
      >
        Deceased Details
      </h1>
    </div>
  </div>
  <div
    className="p-a300 b--w"
  />
</div>
`;

exports[`rendering renders a certificate 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/death/checkout"
        onClick={[Function]}
      >
        Checkout
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/death/checkout"
        onClick={[Function]}
      >
        0
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0"
    >
      <h1
        className="sh-title"
        style={
          Object {
            "marginBottom": 0,
          }
        }
      >
        Deceased Details
      </h1>
    </div>
  </div>
  <div
    className="p-a300 b--w"
  >
    <div
      data-jsx={3255886522}
    >
      <ul
        className="dl"
        data-jsx={3255886522}
      >
        <li
          className="dl-i"
          data-jsx={3255886522}
        >
          <span
            className="dl-t"
            data-jsx={3255886522}
          >
            Full Name
          </span>
          <span
            className="dl-d"
            data-jsx={3255886522}
          >
            Bruce
             
            Banner
          </span>
        </li>
        <li
          className="dl-i"
          data-jsx={3255886522}
        >
          <span
            className="dl-t"
            data-jsx={3255886522}
          >
            Date of Death
          </span>
          <span
            className="dl-d"
            data-jsx={3255886522}
          >
            3/4/2016
          </span>
        </li>
        <li
          className="dl-i"
          data-jsx={3255886522}
        >
          <span
            className="dl-t"
            data-jsx={3255886522}
          >
            Age
          </span>
          <span
            className="dl-d"
            data-jsx={3255886522}
          >
            21 yrs.
          </span>
        </li>
      </ul>
      <form
        className="js-add-to-cart-form m-v300"
        data-jsx={3255886522}
        onSubmit={[Function]}
      >
        <select
          className="quantity"
          data-jsx={3255886522}
          name="quantity"
          onChange={[Function]}
          value={1}
        >
          <option
            data-jsx={3255886522}
            value="1"
          >
            Qty: 1
          </option>
          <option
            data-jsx={3255886522}
            value="2"
          >
            Qty: 2
          </option>
          <option
            data-jsx={3255886522}
            value="3"
          >
            Qty: 3
          </option>
          <option
            data-jsx={3255886522}
            value="4"
          >
            Qty: 4
          </option>
          <option
            data-jsx={3255886522}
            value="5"
          >
            Qty: 5
          </option>
          <option
            data-jsx={3255886522}
            value="6"
          >
            Qty: 6
          </option>
          <option
            data-jsx={3255886522}
            value="7"
          >
            Qty: 7
          </option>
          <option
            data-jsx={3255886522}
            value="8"
          >
            Qty: 8
          </option>
          <option
            data-jsx={3255886522}
            value="9"
          >
            Qty: 9
          </option>
          <option
            data-jsx={3255886522}
            value="10"
          >
            Qty: 10
          </option>
        </select>
        <button
          className="btn add-to-cart"
          data-jsx={3255886522}
          type="submit"
        >
          Add to Cart
        </button>
      </form>
    </div>
  </div>
</div>
`;
